### PR TITLE
Add supports for :links and :data options to has_many associations for jsonapi adapter

### LINKED ISSUE #834: JSONAPI resource relation urls
Currently, the jsonapi adapter adds resource relationships as link objects with ids and types. The jsonapi spec also allows the the resource relationship to be defined as a URL (see http://jsonapi.org/format/#document-structure-resource-relationships). For example:

```json
{
  "type": "articles",
  "id": "1",
  "links": {
    "comments": "http://example.com/articles/1/comments"
  }
}
```
There is currently no way to do this using ActiveModel::Serializers as far as I can tell. 

I propose that there should be a `:url` option for the assoications methods `has_many` and `belongs_to` that can be used to generate an resource url instead of linked objects, so that the above resource could be generated by the following serializer:

```ruby
class ArticleSerializer
  has_many :comments, url: [:article, :comments]
end
```

I am not sure if this option is useful for the other serializers, so feedback on this would be appreciated.

If you are happy with this suggestion I can work on the implementation for this.

### PR DESCRIPTION
**Pull request by lsylvester**
Adds option to `has_many` relationships have links (`links: true`) and turn of data (`data: false`) for jsonapi.

Replaces #840 
Fixes #834

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Hey @lsylvester, I want to review this one, see cool! Could you rebase it please? :smile:

**Comment by lsylvester** [comment]
@joaomdmoura Rebased.

**Comment by kimsuelim** [comment]
@lsylvester, How does this work for namespace url?

For example:
`/api/posts/1/comments`, `api_post_comments(post)`

File: test/fixtures/poro.rb [hunk 1]
```diff
@@ -33,6 +38,10 @@ def id
     @attributes[:id] || @attributes['id'] || object_id
   end
 
+  def to_param
+    id.to_s
+  end
```
Comment (bf4) [inline test/fixtures/poro.rb]: Hiya, Thanks for this.  Would you mind helping me understand these changes?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 2]
```diff
@@ -41,10 +43,17 @@ def fragment_cache(cached_hash, non_cached_hash)
...
-          resource[:relationships][name] ||= { data: [] }
-          resource[:relationships][name][:data] += serializers.map { |serializer| { type: serializer.type, id: serializer.id.to_s } }
+
+          if opts.fetch(:data, true)
+            resource[:relationships][name] ||= { data: [] }
+            resource[:relationships][name][:data] += serializers.map { |serializer| { type: serializer.type, id: serializer.id.to_s } }
+          end
+          if opts.fetch(:links, false)
+            resource[:relationships][name] ||= {}
+            resource[:relationships][name][:links] = { related: url_helper.url_for([serializer.object, name]) }
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Did you consider using `Rails.application.routes.url_helpers` here instead of a reference to the controller?

File: test/fixtures/poro.rb (hunk 1 above)
Comment (joaomdmoura) [inline test/fixtures/poro.rb]: Yup, could you help us on understanding these? :smile:

**Comment by bf4** [comment]
@lsylvester You still working on this?

File: test/fixtures/poro.rb (hunk 1 above)
Comment (lsylvester) [inline test/fixtures/poro.rb]: `to_param` and `to_model` are used in the for url generation.

`to_model` is called by `polymorphic_url` and it is expected to return an ActionModel compliant model. Without it, the `polymororphic_url`v method will error.
`to_param` is used to generate the value for the object in the url. Without it, I get a url like `http://test.host/tags/%23%3CTag:0xXXXXXX%3E/posts\`

File: lib/active_model/serializer/adapter/json_api.rb (hunk 2 above)
Comment (lsylvester) [inline lib/active_model/serializer/adapter/json_api.rb]: `Rails.application.routes.url_helpers` doesn't provide a `polymorphic_url` method, and at the time I wasn't able to work out how to access `polymorphic_url` without a controller. If you think that it is worth removing the need to have a controller I can have another go at it.

**Comment by lsylvester** [comment]
@kimsuelim The is currently no way to provide a name spaced url. Should this be a config option like

```ruby
ActiveModel::Serializer.config.url_namespace = :api
```

or 

```ruby
class SomeSerializer < ActiveModel::Serializer
  url_namespace :api
end
```

File: lib/active_model/serializer/adapter/json_api.rb [hunk 3]
```diff
@@ -44,10 +46,17 @@ def fragment_cache(cached_hash, non_cached_hash)
...
-          resource[:relationships][name] ||= { data: [] }
-          resource[:relationships][name][:data] += serializers.map { |serializer| { type: serializer.json_api_type, id: serializer.id.to_s } }
+
+          if opts.fetch(:data, true)
+            resource[:relationships][name] ||= { data: [] }
+            resource[:relationships][name][:data] += serializers.map { |serializer| { type: serializer.json_api_type, id: serializer.id.to_s } }
+          end
+          if opts.fetch(:links, false)
+            resource[:relationships][name] ||= {}
+            resource[:relationships][name][:links] = { related: url_helper.url_for([serializer.object, name]) }
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Might this PR want to talk to  #1018 (though these are different links)

File: lib/action_controller/serialization.rb [hunk 4]
```diff
@@ -27,6 +27,7 @@ def get_serializer(resource, options = {})
       end
       ActiveModel::SerializableResource.serialize(resource, options) do |serializable_resource|
         if serializable_resource.serializer?
+          serializable_resource.url_helper = self
```
Comment (bf4) [inline lib/action_controller/serialization.rb]: makes me sad we need to do this :( so much coupling, but I guess that's where we're going.  I coulda sworn you could

```ruby
url = Rails.application.routes.url_for(:controller => "users",

                                           :action => "verify_email",

                                           :id_token => "tok",

                                           :only_path => true)
```

Comment (lsylvester) [inline lib/action_controller/serialization.rb]: You can do that, but what we are trying to is more like `Rails.application.routes.url_for([@post, :comments])`

Comment (bf4) [inline lib/action_controller/serialization.rb]: That's right. Any way we can at least narrow the interface to not include the entire controller? This is just be being paranoid of scope creep enabled by it being easy to do.  I like it when tools make it hard to do things the authors don't want you to :)

File: lib/active_model/serializer/adapter/json_api.rb (hunk 3 above)
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: #1018 is kind stopped for sometime, I wouldn't worry abou that right now, but indeed:

> So the link building code should ideally be reusable in different contexts

This would be nice.

File: test/adapter/json_api/has_many_url_test.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            relationships: {
+              posts: {
+                links: {
+                  related: "http://test.host/tags/1/posts"
+                }
+              }
+            }
+          }
+        }
+        assert_equal expected.to_json, response.body
```
Comment (joaomdmoura) [inline test/adapter/json_api/has_many_url_test.rb]: Would be nice to have test covering with both `data` and `links` as true

File: test/fixtures/poro.rb (hunk 1 above)
Comment (joaomdmoura) [inline test/fixtures/poro.rb]: :+1:  Nice, could you add some comments on top of those method explaining this to help new contributors when jumping into the code :smile:

**Comment by joaomdmoura** [comment]
This is really nice @lsylvester great work, I'm looking forward to see this merged.
I made some small comments other the that, this PR looks good.

I would like to ask to also add some docs about this on our [new docs](https://github.com/rails-api/active_model_serializers/tree/master/docs), maybe a new article  on `how to` section. 
btw I know it also represents some more dependencies on Rails and that's the way things are going, but would be nice to have on the docs some work arounds ppl using AMS on PORO could do in order to be able to make it work.

Sorry about this again :grin:, could you rebase it?

**Comment by tchak** [comment]
@lsylvester are you still actively working on this?

It would be also nice to be able to pass an arbitrary object with links

**Comment by tchak** [comment]
I rebased this PR in #1282 and added more tests

**Comment by lsylvester** [comment]
@tchak I am happy for you to take over this one. Closing in favour of #1282
